handler/program.go

```diff
@@ -30,3 +34,86 @@ func GetProgram(cc echo.Context) error {
 
 	return c.JSON(http.StatusOK, &p)
 }
+
+func CreateProgram(cc echo.Context) error {
+	var requestBody struct {
+		UID  string     `json:"uid"`
+		WID  string     `json:"wid"`
+		Prog db.Program `json:"program"`
+	}
+
+	c := cc.(*db.DBContext)
+
+	if err := httpext.RequestBodyTo(c.Request(), &requestBody); err != nil {
+		return c.String(http.StatusInternalServerError, errors.Wrap(err, "failed to read request body").Error())
+	}
+
+	// check that language exists.
+	p := db.DefaultProgram(requestBody.Prog.Language)
+	if p.Code == "" {
+		return c.String(http.StatusBadRequest, "language does not exist")
+	}
+
+	// thumbnail should be within range.
+	if requestBody.Prog.Thumbnail > db.ThumbnailCount || requestBody.Prog.Thumbnail < 0 {
+		return c.String(http.StatusBadRequest, "thumbnail index out of bounds")
+	}
+	p.Thumbnail = requestBody.Prog.Thumbnail
+
+	// add code if provided.
+	if requestBody.Prog.Code != "" {
+		p.Code = requestBody.Prog.Code
+	}
+
+	// add name if provided.
+	if requestBody.Prog.Name != "" {
+		p.Name = requestBody.Prog.Name
+	}
+
+	wid := requestBody.WID
+	var cid string
+	var class db.Class
+	var err error
+
+	if wid != "" {
+		cid, err = c.GetUIDFromWID(c.Request().Context(), wid, db.ClassesAliasPath)
+		if err != nil {
+			return err
+		}
+
+		class, err = c.LoadClass(c.Request().Context(), cid)
+		if err != nil {
+			return err
+		}
+	}
+
+	// create program
+	pRef, _ := c.CreateProgram(c.Request().Context(), requestBody.Prog)
+
+	// associate to user, if they exist
+	u, _ := c.LoadUser(c.Request().Context(), requestBody.UID)
```

Comment:
The _ should be replaced with `err`, u needs to be defined before and `:=` should be replaced with `=`